source/extensions/transport_sockets/alts/tsi_socket.cc

```diff
@@ -317,19 +331,16 @@ Network::IoResult TsiSocket::doWrite(Buffer::Instance& buffer, bool end_stream)
     Network::PostIoAction action = doHandshake();
     ASSERT(action == Network::PostIoAction::KeepOpen);
     // TODO(lizan): Handle synchronous handshake when TsiHandshaker supports it.
-  }
-
-  if (handshake_complete_) {
+    if (raw_write_buffer_.length() > 0) {
+      ENVOY_CONN_LOG(debug, "TSI: raw_write length {} end_stream {}", callbacks_->connection(),
+                     raw_write_buffer_.length(), end_stream);
+      return raw_buffer_socket_->doWrite(raw_write_buffer_, end_stream && (buffer.length() == 0));
```

Comment:
No bytes from buffer were consumed so bytes written should be reported as being 0 regardless of what doWrite on the raw socket says.

Also, in cases where handshake completes successfully, it would be ideal to fall through and execute repeatProtectAndWrite

Reply:
Good point. Done. 

Regarding fall through, I do not think it is going to happen because ALTS TSI is an async implementation which means in doHandshake(), it calls next() and immediately returns, which leaves handshake_complete_ being still false. When the response from handshake service comes back, handshake_complete_ will be set to true if handshake completes successfully.

Reply:
Can you add ASSERT(!handshake_complete_) after the doHandshake() call with an explanation?

Reply:
Done.